File: mobile/lib/core/models/nutrition.dart
/// Modèles Nutrition — SOMA LOT 6.
///
/// FoodItem, NutritionEntry, DailyNutritionSummary, NutritionPhoto.
library;

// ── Aliment ───────────────────────────────────────────────────────────────────

class FoodItem {
  final String id;
  final String name;
  final String? nameFr;
  final double caloriesPer100g;
  final double proteinGPer100g;
  final double carbsGPer100g;
  final double fatGPer100g;
  final double? fiberGPer100g;
  final String? foodGroup;
  final String? source;

  const FoodItem({
    required this.id,
    required this.name,
    this.nameFr,
    required this.caloriesPer100g,
    required this.proteinGPer100g,
    required this.carbsGPer100g,
    required this.fatGPer100g,
    this.fiberGPer100g,
    this.foodGroup,
    this.source,
  });

  factory FoodItem.fromJson(Map<String, dynamic> json) => FoodItem(
        id: json['id'] as String,
        name: json['name'] as String,
        nameFr: json['name_fr'] as String?,
        caloriesPer100g: (json['calories_per_100g'] as num).toDouble(),
        proteinGPer100g: (json['protein_g_per_100g'] as num).toDouble(),
        carbsGPer100g: (json['carbs_g_per_100g'] as num).toDouble(),
        fatGPer100g: (json['fat_g_per_100g'] as num).toDouble(),
        fiberGPer100g: (json['fiber_g_per_100g'] as num?)?.toDouble(),
        foodGroup: json['food_group'] as String?,
        source: json['source'] as String?,
      );

  String get displayName => nameFr ?? name;

  /// Macros pour une quantité donnée en grammes.
  double calories(double quantityG) => caloriesPer100g * quantityG / 100;
  double protein(double quantityG) => proteinGPer100g * quantityG / 100;
  double carbs(double quantityG) => carbsGPer100g * quantityG / 100;
  double fat(double quantityG) => fatGPer100g * quantityG / 100;
}

// ── Entrée nutrition ──────────────────────────────────────────────────────────

class NutritionEntry {
  final String id;
  final String loggedAt;
  final String mealType;
  final String? mealName;
  final double? calories;
  final double? proteinG;
  final double? carbsG;
  final double? fatG;
  final double? fiberG;
  final double? quantityG;
  final String? foodItemId;
  final String? foodItemName;
  final String dataQuality;
  final String? notes;

  const NutritionEntry({
    required this.id,
    required this.loggedAt,
    required this.mealType,
    this.mealName,
    this.calories,
    this.proteinG,
    this.carbsG,
    this.fatG,
    this.fiberG,
    this.quantityG,
    this.foodItemId,
    this.foodItemName,
    required this.dataQuality,
    this.notes,
  });

  factory NutritionEntry.fromJson(Map<String, dynamic> json) => NutritionEntry(
        id: json['id'] as String,
        loggedAt: json['logged_at'] as String,
        mealType: json['meal_type'] as String,
        mealName: json['meal_name'] as String?,
        calories: (json['calories'] as num?)?.toDouble(),
        proteinG: (json['protein_g'] as num?)?.toDouble(),
        carbsG: (json['carbs_g'] as num?)?.toDouble(),
        fatG: (json['fat_g'] as num?)?.toDouble(),
        fiberG: (json['fiber_g'] as num?)?.toDouble(),
        quantityG: (json['quantity_g'] as num?)?.toDouble(),
        foodItemId: json['food_item_id'] as String?,
        foodItemName: json['food_item_name'] as String?,
        dataQuality: json['data_quality'] as String? ?? 'estimated',
        notes: json['notes'] as String?,
      );

  String get mealTypeLabel {
    switch (mealType) {
      case 'breakfast':
        return 'Petit-déjeuner';
      case 'lunch':
        return 'Déjeuner';
      case 'dinner':
        return 'Dîner';
      case 'snack':
        return 'En-cas';
      default:
        return mealType;
    }
  }
}

// ── Résumé nutrition ──────────────────────────────────────────────────────────

class MacroTotals {
  final double calories;
  final double proteinG;
  final double carbsG;
  final double fatG;
  final double fiberG;

  const MacroTotals({
    required this.calories,
    required this.proteinG,
    required this.carbsG,
    required this.fatG,
    required this.fiberG,
  });

  factory MacroTotals.fromJson(Map<String, dynamic> json) => MacroTotals(
        calories: (json['calories'] as num?)?.toDouble() ?? 0,
        proteinG: (json['protein_g'] as num?)?.toDouble() ?? 0,
        carbsG: (json['carbs_g'] as num?)?.toDouble() ?? 0,
        fatG: (json['fat_g'] as num?)?.toDouble() ?? 0,
        fiberG: (json['fiber_g'] as num?)?.toDouble() ?? 0,
      );
}

class MacroGoals {
  final double? caloriesTarget;
  final double? proteinTargetG;

  const MacroGoals({this.caloriesTarget, this.proteinTargetG});

  factory MacroGoals.fromJson(Map<String, dynamic> json) => MacroGoals(
        caloriesTarget: (json['calories_target'] as num?)?.toDouble(),
        proteinTargetG: (json['protein_target_g'] as num?)?.toDouble(),
      );
}

class DailyNutritionSummary {
  final String date;
  final int mealCount;
  final MacroTotals totals;
  final MacroGoals goals;
  final List<NutritionEntry> meals;
  final double dataCompletenessPct;

  const DailyNutritionSummary({
    required this.date,
    required this.mealCount,
    required this.totals,
    required this.goals,
    required this.meals,
    required this.dataCompletenessPct,
  });

  factory DailyNutritionSummary.fromJson(Map<String, dynamic> json) {
    final mealsRaw = json['meals'] as List<dynamic>? ?? [];
    return DailyNutritionSummary(
      date: json['date'] as String,
      mealCount: json['meal_count'] as int? ?? 0,
      totals: MacroTotals.fromJson(
          json['totals'] as Map<String, dynamic>? ?? {}),
      goals: MacroGoals.fromJson(
          json['goals'] as Map<String, dynamic>? ?? {}),
      meals: mealsRaw
          .map((e) => NutritionEntry.fromJson(e as Map<String, dynamic>))
          .toList(),
      dataCompletenessPct:
          (json['data_completeness_pct'] as num?)?.toDouble() ?? 0,
    );
  }

  double get caloriePct {
    final t = goals.caloriesTarget;
    if (t == null || t == 0) return 0;
    return (totals.calories / t * 100).clamp(0, 100);
  }

  double get proteinPct {
    final t = goals.proteinTargetG;
    if (t == null || t == 0) return 0;
    return (totals.proteinG / t * 100).clamp(0, 100);
  }
}

// ── Photo repas ───────────────────────────────────────────────────────────────

class DetectedFood {
  final String name;
  final double? quantityG;
  final double? caloriesEstimated;
  final double? confidence;

  const DetectedFood({
    required this.name,
    this.quantityG,
    this.caloriesEstimated,
    this.confidence,
  });

  factory DetectedFood.fromJson(Map<String, dynamic> json) => DetectedFood(
        name: json['name'] as String,
        quantityG: (json['quantity_g'] as num?)?.toDouble(),
        caloriesEstimated: (json['calories_estimated'] as num?)?.toDouble(),
        confidence: (json['confidence'] as num?)?.toDouble(),
      );
}

class NutritionPhoto {
  final String photoId;
  final String analysisStatus; // pending|analyzing|analyzed|failed
  final List<DetectedFood> identifiedFoods;
  final double? estimatedCalories;
  final double? estimatedProteinG;
  final double? estimatedCarbsG;
  final double? estimatedFatG;
  final double? overallConfidence;
  final String? mealTypeGuess;

  const NutritionPhoto({
    required this.photoId,
    required this.analysisStatus,
    required this.identifiedFoods,
    this.estimatedCalories,
    this.estimatedProteinG,
    this.estimatedCarbsG,
    this.estimatedFatG,
    this.overallConfidence,
    this.mealTypeGuess,
  });

  factory NutritionPhoto.fromJson(Map<String, dynamic> json) {
    final foods = json['identified_foods'] as List<dynamic>? ?? [];
    return NutritionPhoto(
      photoId: json['photo_id'] as String,
      analysisStatus: json['analysis_status'] as String? ?? 'pending',
      identifiedFoods: foods
          .map((e) => DetectedFood.fromJson(e as Map<String, dynamic>))
          .toList(),
      estimatedCalories: (json['estimated_calories'] as num?)?.toDouble(),
      estimatedProteinG: (json['estimated_protein_g'] as num?)?.toDouble(),
      estimatedCarbsG: (json['estimated_carbs_g'] as num?)?.toDouble(),
      estimatedFatG: (json['estimated_fat_g'] as num?)?.toDouble(),
      overallConfidence: (json['overall_confidence'] as num?)?.toDouble(),
      mealTypeGuess: json['meal_type_guess'] as String?,
    );
  }

  bool get isAnalyzed => analysisStatus == 'analyzed';
  bool get isFailed => analysisStatus == 'failed';
  bool get isPending =>
      analysisStatus == 'pending' || analysisStatus == 'analyzing';
}

File: backend/app/schemas/nutrition.py
"""Schémas Pydantic v2 — module nutrition (LOT 2)."""
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field, model_validator
import uuid


# ── Food Items ─────────────────────────────────────────────────────────────────

class FoodItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    name_fr: Optional[str] = None
    barcode: Optional[str] = None
    # Macros / 100g
    calories_per_100g: Optional[float] = None
    protein_g_per_100g: Optional[float] = None
    carbs_g_per_100g: Optional[float] = None
    fat_g_per_100g: Optional[float] = None
    fiber_g_per_100g: Optional[float] = None
    sugar_g_per_100g: Optional[float] = None
    # Classification
    food_group: Optional[str] = None
    is_ultra_processed: bool = False
    nova_score: Optional[int] = None
    source: Optional[str] = None
    verified: bool = False


class FoodItemListResponse(BaseModel):
    items: List[FoodItemResponse]
    total: int
    page: int
    per_page: int


# ── Nutrition Entries ───────────────────────────────────────────────────────────

class NutritionEntryCreate(BaseModel):
    """
    Création d'une entrée nutritionnelle.

    Trois modes de saisie :
    A. Via food_item_id + quantity_g → macros auto-calculées si non fournies
    B. Via photo_id → macros pré-remplies depuis l'analyse photo
    C. Saisie directe des macros (estimation manuelle)
    """
    logged_at: Optional[datetime] = None        # défaut = now()
    meal_type: Optional[str] = Field(
        None,
        pattern="^(breakfast|lunch|dinner|snack|supplement|drink)$",
    )
    meal_name: Optional[str] = Field(None, max_length=200)

    # Source
    food_item_id: Optional[uuid.UUID] = None
    photo_id: Optional[uuid.UUID] = None
    quantity_g: Optional[float] = Field(None, ge=0.0)

    # Macros (explicites ou auto-calculées)
    calories: Optional[float] = Field(None, ge=0.0)
    protein_g: Optional[float] = Field(None, ge=0.0)
    carbs_g: Optional[float] = Field(None, ge=0.0)
    fat_g: Optional[float] = Field(None, ge=0.0)
    fiber_g: Optional[float] = Field(None, ge=0.0)

    # Qualité de la donnée
    data_quality: Optional[str] = Field(
        None, pattern="^(exact|estimated|inferred|ai_analyzed)$"
    )

    # Contexte
    hunger_before: Optional[int] = Field(None, ge=1, le=10)
    satiety_after: Optional[int] = Field(None, ge=1, le=10)
    energy_after: Optional[int] = Field(None, ge=1, le=10)
    notes: Optional[str] = None
    fasting_window_broken: bool = False

    @model_validator(mode="after")
    def validate_at_least_one_source(self) -> "NutritionEntryCreate":
        """Au moins une source ou des macros explicites sont nécessaires."""
        has_source = bool(self.food_item_id or self.photo_id)
        has_macros = any([self.calories, self.protein_g, self.carbs_g, self.fat_g])
        if not has_source and not has_macros:
            raise ValueError(
                "Au moins un food_item_id, un photo_id, ou des macros explicites sont nécessaires."
            )
        return self


class NutritionEntryUpdate(BaseModel):
    meal_type: Optional[str] = Field(None, pattern="^(breakfast|lunch|dinner|snack|supplement|drink)$")
    meal_name: Optional[str] = Field(None, max_length=200)
    logged_at: Optional[datetime] = None
    quantity_g: Optional[float] = Field(None, ge=0.0)
    calories: Optional[float] = Field(None, ge=0.0)
    protein_g: Optional[float] = Field(None, ge=0.0)
    carbs_g: Optional[float] = Field(None, ge=0.0)
    fat_g: Optional[float] = Field(None, ge=0.0)
    fiber_g: Optional[float] = Field(None, ge=0.0)
    data_quality: Optional[str] = Field(None, pattern="^(exact|estimated|inferred|ai_analyzed)$")
    hunger_before: Optional[int] = Field(None, ge=1, le=10)
    satiety_after: Optional[int] = Field(None, ge=1, le=10)
    energy_after: Optional[int] = Field(None, ge=1, le=10)
    notes: Optional[str] = None


class NutritionEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    logged_at: datetime
    meal_type: Optional[str] = None
    meal_name: Optional[str] = None
    food_item_id: Optional[uuid.UUID] = None
    photo_id: Optional[uuid.UUID] = None
    quantity_g: Optional[float] = None
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    carbs_g: Optional[float] = None
    fat_g: Optional[float] = None
    fiber_g: Optional[float] = None
    data_quality: Optional[str] = None
    hunger_before: Optional[int] = None
    satiety_after: Optional[int] = None
    energy_after: Optional[int] = None
    notes: Optional[str] = None
    fasting_window_broken: bool = False
    created_at: datetime
    updated_at: datetime


class NutritionEntryListResponse(BaseModel):
    entries: List[NutritionEntryResponse]
    total: int
    date: Optional[str] = None


# ── Daily Nutrition Summary ────────────────────────────────────────────────────

class MacroGoals(BaseModel):
    """Objectifs nutritionnels du profil utilisateur."""
    calories_target: Optional[float] = None
    protein_target_g: Optional[float] = None
    carbs_target_g: Optional[float] = None
    fat_target_g: Optional[float] = None


class MacroActuals(BaseModel):
    """Totaux nutritionnels réels du jour."""
    calories: float = 0.0
    protein_g: float = 0.0
    carbs_g: float = 0.0
    fat_g: float = 0.0
    fiber_g: float = 0.0


class MacroBalance(BaseModel):
    """Écart entre réel et objectif (positif = surplus, négatif = déficit)."""
    calories_delta: Optional[float] = None
    protein_delta_g: Optional[float] = None
    carbs_delta_g: Optional[float] = None
    fat_delta_g: Optional[float] = None
    pct_calories_reached: Optional[float] = None
    pct_protein_reached: Optional[float] = None


class MealSummaryItem(BaseModel):
    """Résumé d'un repas dans le journal."""
    id: uuid.UUID
    meal_type: Optional[str] = None
    meal_name: Optional[str] = None
    logged_at: datetime
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    data_quality: Optional[str] = None


class EatingWindow(BaseModel):
    """Fenêtre alimentaire estimée (premier repas → dernier repas)."""
    first_meal_at: Optional[datetime] = None
    last_meal_at: Optional[datetime] = None
    window_hours: Optional[float] = None
    fasting_compatible: Optional[bool] = None  # window ≤ 8h si jeûne 16:8


class DailyNutritionSummary(BaseModel):
    date: str
    meal_count: int
    totals: MacroActuals
    goals: Optional[MacroGoals] = None
    balance: Optional[MacroBalance] = None
    eating_window: EatingWindow
    meals: List[MealSummaryItem] = []
    data_completeness_pct: float     # % d'entrées avec calories renseignées
    has_photo_entries: bool = False


# ── Photo Upload & Analysis Pipeline ──────────────────────────────────────────

class NutritionPhotoUploadResponse(BaseModel):
    """Réponse immédiate après upload (analyse asynchrone)."""
    photo_id: uuid.UUID
    status: str  # "pending" ou "analyzing"
    message: str


class DetectedFoodItem(BaseModel):
    """Aliment détecté par Claude Vision."""
    name: str
    name_fr: Optional[str] = None
    quantity_g: Optional[float] = None
    calories_estimated: Optional[float] = None
    protein_g_estimated: Optional[float] = None
    carbs_g_estimated: Optional[float] = None
    fat_g_estimated: Optional[float] = None
    confidence: float = Field(..., ge=0.0, le=1.0)
    food_group: Optional[str] = None
    notes: Optional[str] = None


class PhotoAnalysisResult(BaseModel):
    """Résultat structuré de l'analyse IA."""
    photo_id: uuid.UUID
    analysis_status: str       # pending, analyzing, analyzed, failed
    identified_foods: Optional[List[DetectedFoodItem]] = None
    estimated_calories: Optional[float] = None
    estimated_protein_g: Optional[float] = None
    estimated_carbs_g: Optional[float] = None
    estimated_fat_g: Optional[float] = None
    overall_confidence: Optional[float] = None
    meal_type_guess: Optional[str] = None
    warnings: Optional[List[str]] = None
    assumptions: Optional[List[str]] = None
    error_message: Optional[str] = None
    created_at: Optional[datetime] = None
    analyzed_at: Optional[datetime] = None


class PhotoConfirmRequest(BaseModel):
    """
    Confirmation par l'utilisateur de l'analyse photo.
    Peut inclure des corrections sur les aliments ou les macros.
    """
    meal_type: Optional[str] = Field(None, pattern="^(breakfast|lunch|dinner|snack|supplement|drink)$")
    meal_name: Optional[str] = Field(None, max_length=200)
    # Corrections optionnelles (remplacent les valeurs IA)
    corrected_calories: Optional[float] = Field(None, ge=0.0)
    corrected_protein_g: Optional[float] = Field(None, ge=0.0)
    corrected_carbs_g: Optional[float] = Field(None, ge=0.0)
    corrected_fat_g: Optional[float] = Field(None, ge=0.0)
    corrected_foods: Optional[List[DetectedFoodItem]] = None
    notes: Optional[str] = None
    # Si True, crée automatiquement une NutritionEntry après confirmation
    create_entry: bool = True


class PhotoConfirmResponse(BaseModel):
    photo_id: uuid.UUID
    user_validated: bool
    meal_type: Optional[str]
    # Si create_entry=True
    entry_id: Optional[uuid.UUID] = None
    final_calories: Optional[float] = None
    final_protein_g: Optional[float] = None
    final_carbs_g: Optional[float] = None
    final_fat_g: Optional[float] = None
